src/ClientTransactionTask.cc

```diff
@@ -0,0 +1,562 @@
+            prepareRpcs.emplace_back(ramcloud, rpcSession, this);
+            nextRpc = &prepareRpcs.back();
+        }
+
+        Transport::SessionRef session =
+                ramcloud->objectFinder.lookup(key->tableId, key->keyHash);
+        if (session->getServiceLocator() == rpcSession->getServiceLocator()
+            && nextRpc->reqHdr->opCount < PrepareRpc::MAX_OBJECTS_PER_RPC) {
+            nextRpc->appendOp(nextCacheEntry);
+        } else {
+            break;
+        }
+    }
+    if (nextRpc) {
+        nextRpc->send();
+    }
+}
+
+/**
+ * Constructor for a decision rpc.
+ *
+ * \param ramcloud
+ *      The RAMCloud object that governs this RPC.
+ * \param session
+ *      Session on which this RPC will eventually be sent.
+ * \param task
+ *      Pointer to the transaction task that issued this request.
+ */
+ClientTransactionTask::DecisionRpc::DecisionRpc(RamCloud* ramcloud,
+        Transport::SessionRef session,
+        ClientTransactionTask* task)
+    : RpcWrapper(sizeof(WireFormat::TxDecision::Response))
+    , ramcloud(ramcloud)
+    , task(task)
+    , ops()
+    , reqHdr(allocHeader<WireFormat::TxDecision>())
+{
+    reqHdr->decision = task->decision;
+    reqHdr->leaseId = task->lease.leaseId;
+    reqHdr->participantCount = 0;
+    this->session = session;
+}
+
+// See RpcWrapper for documentation.
+bool
+ClientTransactionTask::DecisionRpc::checkStatus()
+{
+    if (responseHeader->status == STATUS_UNKNOWN_TABLET) {
+        retryRequest();
+    }
+    return true;
+}
+
+// See RpcWrapper for documentation.
+bool
+ClientTransactionTask::DecisionRpc::handleTransportError()
+{
+    // There was a transport-level failure. Flush cached state related
+    // to this session, and related to the object mappings.  The objects
+    // will all be retried when \c finish is called.
+    if (session.get() != NULL) {
+        ramcloud->clientContext->transportManager->flushSession(
+                session->getServiceLocator());
+        session = NULL;
+    }
+    retryRequest();
+    return true;
+}
+
+// See RpcWrapper for documentation.
+void
+ClientTransactionTask::DecisionRpc::send()
+{
+    state = IN_PROGRESS;
+    session->sendRequest(&request, response, this);
+}
+
+/**
+ * Append an operation to the end of this decision rpc.
+ *
+ * \param opEntry
+ *      Handle to information about the operation to be appended.
+ */
+void
+ClientTransactionTask::DecisionRpc::appendOp(CommitCacheMap::iterator opEntry)
+{
+    const CacheKey* key = &opEntry->first;
+    CacheEntry* entry = &opEntry->second;
+
+    request.emplaceAppend<WireFormat::TxParticipant>(
+            key->tableId,
+            static_cast<uint64_t>(key->keyHash),
+            entry->rpcId);
+
+    entry->state = CacheEntry::DECIDE;
+    ops[reqHdr->participantCount] = opEntry;
+    reqHdr->participantCount++;
+}
+
+/**
+ * Handle the case where the RPC may have been sent to the wrong server.
+ */
+void
+ClientTransactionTask::DecisionRpc::retryRequest()
+{
```

Comment:
I think you might be able to combine this method with PrepareRpc::retryRequest; let's discuss.